driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -392,7 +407,7 @@ private void maybeSpawnNewConnection() {
     }
 
     void replaceDefunctConnection(final PooledConnection connection) {
-        if (connection.markForTrash.compareAndSet(false, true))
+        if (connection.state.compareAndSet(OPEN, GONE))
             open.decrementAndGet();
         connections.remove(connection);
         connection.closeAsync();
```

Comment:
Fixed in 4c5a06a. I've used the result of `connections.remove` for the test.

Note that `addConnectionIfUnderMaximum` checks the number of connections anyway, so this is just to avoid scheduling an unnecessary task.